libgadget/run.c

```diff
@@ -271,6 +274,13 @@ run(int RestartSnapNum)
             /* Scratch data cannot be used checkpoint because FOF does an exchange.*/
             slots_free_sph_pred_data(&sph_predicted);
         }
+        
+        if(All.FdmOn)
+        {
+            dm_density(&Act,&Tree);
+            force_update_dm_hmax(Act.ActiveParticle, Act.NumActiveParticle, &Tree, ddecomp);
```

Comment:
Actually I am not sure whether I need to do this force_update_dm_hmax

Reply:
If not, I will concatenate the three QP-treewalks together into one function

Reply:
You only need force_update_dm_hmax for TREEWALK_SYMMETRIC, and all the treewalks as written are TREEWALK_ASYMMETRIC . But are you sure they should be? The hydro pressure is the classic example of TREEWALK_SYMMETRIC. I don't remember from the FDM papers whether these equations need to be symmetric.

Reply:
If Nori 2018 doesn't do it perhaps we shouldn't either indeed. But I'm not sure! I haven't read these papers for a while.